conftest.py

```diff
@@ -0,0 +1,166 @@
+import os
+from uuid import uuid1
+
+import pytest
+
+from hub.constants import (MIN_LOCAL_CACHE_SIZE, MIN_MEMORY_CACHE_SIZE,
+                           PYTEST_LOCAL_PROVIDER_BASE_ROOT,
+                           PYTEST_MEMORY_PROVIDER_BASE_ROOT,
+                           PYTEST_S3_PROVIDER_BASE_ROOT)
+from hub.core.storage import LocalProvider, MemoryProvider, S3Provider
+from hub.tests.common import SESSION_ID, current_test_name
+from hub.util.cache_chain import get_cache_chain
+
+STORAGE_CONFIG_MAP = {
+    "memory": {
+        "base_root": PYTEST_MEMORY_PROVIDER_BASE_ROOT,
+        "class": MemoryProvider,
+        "use_id": False,
+        "is_id_prefix": False,  # if is_id_prefix (and use_id=True), the session id comes before test name, otherwise it is reversed
+    },
+    "local": {
+        "base_root": PYTEST_LOCAL_PROVIDER_BASE_ROOT,
+        "class": LocalProvider,
+        "use_id": False,
+        "is_id_prefix": False,
+    },
+    "s3": {
+        "base_root": PYTEST_S3_PROVIDER_BASE_ROOT,
+        "class": S3Provider,
+        "use_id": True,
+        "is_id_prefix": True,
+    },
+}
+
+
+def _skip_if_none(val):
+    if val is None:
+        pytest.skip()
+
+
+def _is_opt_true(request, opt):
+    return request.config.getoption(opt)
+
+
+def pytest_addoption(parser):
+    parser.addoption(
+        "--memory-skip",
+        action="store_true",
+        help="Tests using the `memory_provider` fixture will be skipped. Tests using the `storage` fixture will be skipped if called with \
+                `MemoryProvider`.",
+    )
+    parser.addoption(
+        "--local",
+        action="store_true",
+        help="Tests using the `storage`/`local_provider` fixtures will run with `LocalProvider`.",
+    )
+    parser.addoption(
+        "--s3",
+        action="store_true",
+        help="Tests using the `storage`/`s3_provider` fixtures will run with `S3Provider`.",
+    )
+    parser.addoption(
+        "--cache-chains",
+        action="store_true",
+        help="Tests using the `storage` fixture may run with combinations of all enabled providers \
+                in cache chains. For example, if the option `--s3` is not provided, all cache chains that use `S3Provider` are skipped.",
+    )
+    parser.addoption(
+        "--cache-chains-only",
+        action="store_true",
+        help="Force enables `--cache-chains`. `storage` fixture only returns cache chains. For example, if `--s3` is provided, \
+            `storage` will never be just `S3Provider`.",
+    )
+
+
+def _get_storage_provider(storage_name, with_current_test_name=True):
+    info = STORAGE_CONFIG_MAP[storage_name]
```

Comment:
Would have preferred to keep the map as a class instead of a string map.

Example: 
class StorageProviderFactory
      String base_root
      Obj S3Provider,
      boolean use_id
      boolean is_id_prefix


def get_local_provider():
	return new config(base_root, s3provider, asas.)


provider_config = get_local_provider()

this-> provider_config.base_root

instead of this-> provider_config["base_root"]

Reply:
not going to do this. i just added functions that wrap `_get_storage_provider` like `_get_memory_provider`. classes here are overkill. a config dictionary is fully acceptable because it's clear how to configure. a class will be more confusing.